File: typescript/src/payment/application/policies/paymentCompletion.ts
// Policy che traduce TransactionAccepted in AcceptTransaction.

import type { Policy } from '@/common/application/policy.js';
import { AcceptTransaction, acceptTransaction } from '../../application/commands/acceptTransaction.js';
import { TransactionAccepted } from '../../domain/events/transactionAccepted.js';
import { requireArgument } from '@/common/utils/requireArgument.js';

export class PaymentCompletion implements Policy<TransactionAccepted, AcceptTransaction> {
  evaluate(event: TransactionAccepted): AcceptTransaction {
    requireArgument(event, 'event');
    return acceptTransaction(event.aggregateId, event.transactionId, event.providerCompletedAt);
  }
}

File: java/src/main/java/it/giannibombelli/wsc2026/payment/application/policies/PaymentCompletion.java
package it.giannibombelli.wsc2026.payment.application.policies;

import it.giannibombelli.wsc2026.common.utils.Require;

import it.giannibombelli.wsc2026.common.application.Policy;
import it.giannibombelli.wsc2026.payment.application.commands.AcceptTransaction;
import it.giannibombelli.wsc2026.payment.domain.events.TransactionAccepted;

public final class PaymentCompletion implements Policy<TransactionAccepted, AcceptTransaction> {

    @Override
    public AcceptTransaction evaluate(TransactionAccepted event) {
        Require.requireArgument(event, "event");
        return new AcceptTransaction(
            event.aggregateId(),
            event.transactionId(),
            event.providerCompletedAt()
        );
    }
}
